### PR TITLE
2.0.0 - January 24, 2023

### PR DESCRIPTION
**Pull request by thpang**
- #25
- #37
- #47
- Update repo about text to mirror new SAS Viya Platform naming
- Addition of the `local-storage` storage class to support:
  - SingleStore
  - OpenSearch
- Load balancer enhancements
  - MetalLB support - Needed for SingleStore load balancer IP assignment
  - More flexible load balancer syntax
- Added metrics-server
- Enhance uninstall - One can now remove kubernetes and re-install as needed
- Kubernetes 1.26 support
  - Includes updated containerd packaging
  - Modifications to kubeadm configuration file for api adjustments.
- Ability to control CNI version
- Removal of auto update task and packages from OS
- Disabling of SWAP
- Several version updates
- Misc.

### CODE REVIEW

File: roles/systems/common/tasks/main.yaml [hunk 1]
```diff
@@ -61,7 +93,38 @@
     curl -skf http://delphi.unx.sas.com/dumpster/smitty2/certs/sasroot2.crt -o /etc/ssl/certs/sasroot2.crt
     curl -skf http://delphi.unx.sas.com/dumpster/smitty2/certs/sasinter.crt -o /etc/ssl/certs/sasinter.crt
     update-ca-certificates
-  when: ansible_distribution == "Ubuntu" and ansible_distribution_version == "20.04"
+  when: ansible_distribution == "Ubuntu" and (ansible_distribution_version == "20.04" or ansible_distribution_version == "22.04")
   tags:
     - install
     - update
+
+- name: Install python and pip packages
```
Comment (supear) [inline roles/systems/common/tasks/main.yaml:101]: Python and PIP

**Review by supear — CHANGES_REQUESTED** [review]
Was in a rush, but I think I caught the most salient issues.

File: README.md [hunk 2]
```diff
@@ -2,29 +2,28 @@
...
-issue is known and is being addressed with [this issue](https://github.com/sassoftware/viya4-iac-k8s/issues/6). As soon as this is resolved this banner will be removed and documentation will be updated to direct you on how to active the correct storage class needed.
+- A problem with the implementation of the default storage class and its usage of an NFS server as its
+backing store has been addressed with [this issue](https://github.com/sassoftware/viya4-iac-k8s/issues/6).
 
 ## Overview
 
-This project contains Terraform scripts to provision cloud infrastructure resources for VMware, and it contains Ansible files to apply the elements of a Kubernetes cluster that are required to deploy SAS Viya 4 product offerings. Here is a list of resources that this project can create:
+This project helps you to automate the cluster-provisioning phase of SAS Viya deployment. It contains Terraform scripts to provision cloud infrastructure resources for VMware, and it contains Ansible files to apply the elements of a Kubernetes cluster that are required to deploy SAS Viya 4 product offerings. Here is a list of resources that this project can create:
```
[the remark this answers is not in the document]
Comment (thpang) [reply README.md:10]: Are you suggesting that the other parts of this paragraph be removed? Not sure about the comment here.

[the remark this answers is not in the document]
Comment (supear) [reply README.md:10]: The change I'm suggesting is just to add the word "platform" for the renaming of SAS Viya to "SAS Viya platform" in that sentence. Everything else can stay the same.

File: roles/kubernetes/loadbalancer/metallb/tasks/main.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+---
+# NOTE: The usage and setting of the iac_inventory_dir variable
+#       is only needed given this task is run on localhost
+#       an implied localhost and this keeps the magic inventory_file
```
[the remark this answers is not in the document]
Comment (thpang) [reply roles/kubernetes/loadbalancer/metallb/tasks/main.yaml:4]: They are needed, but not provided. Hence the wording and your initial correction.

**Comment by thpang** [comment]
@supear I have made updates based on your review comments. If you could take another pass. Thx so very much.

File: README.md [hunk 4]
```diff
@@ -19,9 +19,9 @@ This project helps you to automate the cluster-provisioning phase of SAS Viya de
...
+To learn about all phases and options of the SAS Viya platform deployment process, see [Getting Started with SAS Viya and Open Source Kubernetes](https://go.documentation.sas.com/doc/en/itopscdc/default/itopscon/p1qungdpndaksyn156ng6duptma0.htm?fromDefault=) in _SAS&reg; Viya&reg; Platform Operations_.
 
-Once the resources are provisioned, use the [viya4-deployment](https://github.com/sassoftware/viya4-deployment) project to deploy SAS Viya in your cloud environment. For more information about SAS Viya requirements and documentation for the deployment process, refer to the [SAS Viya Operations Guide](https://go.documentation.sas.com/doc/en/itopscdc/default/itopswn/titlepage.htm).
+Once the resources are provisioned, use the [viya4-deployment](https://github.com/sassoftware/viya4-deployment) project to deploy SAS Viya in platform your cloud environment. For more information about SAS Viya platform requirements and documentation for the deployment process, refer to [SAS Viya Platform Operations](https://go.documentation.sas.com/doc/en/itopscdc/default/itopswn/titlepage.htm).
```
Comment (supear) [inline README.md:24]: SAS Viya in platform -> SAS Viya platform in

**Review by supear — DISMISSED** [review]
Just saw one typo!

**Comment by thpang** [comment]
@supear looking for the typo you reference. If you could point me to its location. Thx.

File: Dockerfile [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
   && update-alternatives --install /usr/bin/python python /usr/bin/python3 1 \
   && update-alternatives --install /usr/bin/pip pip /usr/bin/pip3 1
 
 # Layers used for building/downloading/installing tools
 FROM baseline as tool_builder
-ARG HELM_VERSION=3.8.1
+ARG HELM_VERSION=3.10.0
 ARG KUBECTL_VERSION=1.23.8
-ARG TERRAFORM_VERSION=1.2.0
+ARG TERRAFORM_VERSION=1.3.2
```
Comment (dhoucgitter) [inline Dockerfile:12]: Does oss-k8s require features present in TF 1.3.2 or is that the version you've been using locally? The other providers still reference TF 1.0.0 in their Dockerfiles.

Comment (dhoucgitter) [reply Dockerfile:12]: Just noticed that the Readme calls out TF 1.3.2 as required below.

File: docs/REQUIREMENTS.md [hunk 6]
```diff
@@ -1,119 +1,185 @@
...
 ## Operating System
 
-An Ubuntu Linux operating system is required for the tasks associated with standing up infrastructure using the tools in this repository.
+An Ubuntu Linux operating system is required for the machine that uses the tools in this repository to perform the tasks associated with standing up infrastructure for a SAS Viya deployment.
 
-| OS | Description |
+| Operating System | Description |
 | --- | --- |
-| [Ubuntu 20.04 LTS](https://releases.ubuntu.com/20.04/) | You must have a user account and password with privileges that enable unprompted `sudo`. You must also have a shared "private/public" SSH key pair to use with each system. These are required for the Ansible tools that are described below. |
+| [Ubuntu 22.04 LTS](https://releases.ubuntu.com/22.04/) | You must have a user account and password with privileges that enable unprompted `sudo`. You must also have a shared "private/public" SSH key pair to use with each system. These are required for the Ansible tools that are described below. |
```
Comment (dhoucgitter) [inline docs/REQUIREMENTS.md:34]: The Readme content lists Ubuntu 20.04 LTS and 22.04 LTS as being supported but this file says 22.04 LTS is required. Should 20.04 LTS be listed here also?

File: docs/REQUIREMENTS.md [hunk 7]
```diff
@@ -1,119 +1,185 @@
...
+## Storage
+
+This section outlines how storage is used by the tooling to create the `local-storage` storage class for use in kubernetes. Currently there are two such products that are used in the SAS Viya 4 system that require local storage, these are **OpenSearch** and **SingleStore**. If these products are going to be in use within your cluster you will need to have local storage active.
+
+### Configuration and Restrictions
+
+1. The disks must be represented as empty partitions - `/dev/sdb`, `/dev/sdc`, etc. and attached to the machine or VM.
+2. The empty partitions cannot be formatted or pre-configured. These partitions must simply be attached and in their raw state.
+3. You must supply an empty partition for each local storage entity needed. Currently if your deployment will include any products that utilize **OpenSearch** or **SingleStore**, you'll need need to have these disks available for use.
+4. You will need to updated ANY transformer being used for the **OpenSearch** or **SingleStore** products to mirror these values. Using `local-storage` as the storage class and adjusting the storage size to mirror the partition created for use.
```
Comment (dhoucgitter) [inline docs/REQUIREMENTS.md:119]:
```suggestion
4. You will need to update ANY transformer being used for the **OpenSearch** or **SingleStore** products to mirror these values. Using `local-storage` as the storage class and adjusting the storage size to mirror the partition created for use.
```

**Review by dhoucgitter — APPROVED** [review]

File: docs/CONFIG-VARS.md [hunk 8]
```diff
@@ -251,19 +280,19 @@ nfs_ip        = ""    # Assigned values for static IP addresses
...
 
 **NOTES**:
 
-1. If you set `server_ssl` to on, and you do not define either `server_ssl_cert_file` or `server_ssl_cert_file` the system's default SSL certificate and key will be used instead. By default, on Ubuntu systems we create a copy of those files and name them `ssl-cert-sas-${PG_HOST}.pem` and `ssl-cert-sas-${PG_HOST}.key`
-    * The Ansible tasks will take care of copying the certificate and key from the Postgres VM into your local workspace directory
-2. If you are planning on using the [viya4-deployment repository](https://github.com/sassoftware/viya4-deployment) to perform a Viya deployment where you have [full-stack TLS](https://github.com/sassoftware/viya4-deployment/blob/main/docs/CONFIG-VARS.md#tls), ensure in the viya4-deployment ansible-vars.yaml the `V4_CFG_TLS_TRUSTED_CA_CERTS` variable points to a directory that contains the `server_ssl_cert_file`.
+1. If you enable `server_ssl` without defining either `server_ssl_cert_file` or `server_ssl_cert_file`, the system's default SSL certificate and key are used instead. By default, on Ubuntu systems we create a copy of those files and name them `ssl-cert-sas-${PG_HOST}.pem` and `ssl-cert-sas-${PG_HOST}.key`.
```
Comment (dhoucgitter) [inline docs/CONFIG-VARS.md:293]:
```suggestion
1. If you enable `server_ssl` without defining either `server_ssl_cert_file` or `server_ssl_key_file`, the system's default SSL certificate and key are used instead. By default, on Ubuntu systems we create a copy of those files and name them `ssl-cert-sas-${PG_HOST}.pem` and `ssl-cert-sas-${PG_HOST}.key`.
```

Comment (thpang) [reply docs/CONFIG-VARS.md:293]: Updated and pushing to staging.

**Review by dhoucgitter — APPROVED** [review]

**Review by anthoday — APPROVED** [review]

**Review by dhoucgitter — APPROVED** [review]

**Pull request merged**